Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp

```diff
@@ -361,6 +364,20 @@ static void addParametersShared(const LocalFrame* frame, NetworkResourceLoadPara
 
     if (isMainFrameNavigation)
         parameters.linkPreconnectEarlyHintsEnabled = mainFrame->settings().linkPreconnectEarlyHintsEnabled();
+    else if (RefPtr localMainFrame = dynamicDowncast<LocalFrame>(mainFrame)) {
+        if (auto* document = frame->document(); localMainFrame->document()
+            && isRequestCrossOrigin(&localMainFrame->document()->securityOrigin(), parameters.request.url(), { })
+            && document) {
+
+            std::optional<SecurityOriginData> securityOrigin;
+            if (cachedResource && cachedResource->type() == CachedResource::Type::MainResource)
+                securityOrigin = SecurityOriginData::fromURL(parameters.request.url());
+            parameters.isPermissionPolicyPrivateTokenAllowed = isFeaturePolicyAllowedByDocumentAndAllOwners(FeaturePolicy::Type::PrivateToken, *document, LogFeaturePolicyFailure::No, securityOrigin);
+        }
+
+        if (auto* page = frame && frame->page() ? WebPage::fromCorePage(*frame->page()) : nullptr; page && page->shouldSendPrivateTokenIPCForTesting())
+            page->send(Messages::WebPageProxy::DidAllowPrivateTokensForThirdPartyForTesting(parameters.isPermissionPolicyPrivateTokenAllowed, parameters.request.url()));
```

Comment:
I am not sure this is the right place for this.
Maybe it should be networking process that would send this information to UIProcess at the moment it triggers the network load. NetworkResourceLoader::startNetworkLoad for instance.